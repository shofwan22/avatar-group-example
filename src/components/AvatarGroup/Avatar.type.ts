export type Size = 'xs' | 'sm' | 'md' | 'lg';

export interface User {
  name: string;
  image?: string;
}

export interface AvatarGroupProps {
  maxLength: number;
  size: Size;
  users: User[];
}
