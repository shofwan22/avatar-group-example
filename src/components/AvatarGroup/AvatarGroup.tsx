import { AvatarGroupProps, User } from './Avatar.types';
import { AVATAR_SIZE, TEXT_SIZE } from '../../constants/avatar';

const AvatarGroup = ({ users, maxLength, size = 'md' }: AvatarGroupProps) => {
  const avatarSize = AVATAR_SIZE[size];
  const textSize = TEXT_SIZE[size];
  const overflowCount = users.length - maxLength;

  const getInitials = (name: string) => {
    const nameParts = name.split(' ');
    const initials =
      nameParts[0].charAt(0).toUpperCase() +
      (nameParts[1] ? nameParts[1].charAt(0).toUpperCase() : '');
    return initials;
  };

  return (
    <div
      data-testid="avatar-container"
      className="flex items-center -space-x-4"
    >
      {users.slice(0, maxLength).map((user: User, idx: number) => (
        <div key={idx} data-testid="avatar-component">
          {user.image ? (
            <img
              className={`${avatarSize} rounded-full`}
              src={user.image}
              alt="avatar"
              data-testid="avatar-image"
            />
          ) : (
            <div
              data-testid="avatar-name"
              className={`flex-shrink-0 ${avatarSize} rounded-full border-2 border-white bg-green-400 text-white ${textSize} font-semibold flex items-center justify-center`}
            >
              {getInitials(user.name)}
            </div>
          )}
        </div>
      ))}
      {overflowCount > 0 && (
        <div
          data-testid="avatar-more"
          className={`flex-shrink-0 ${avatarSize} rounded-full border-2 border-white bg-slate-300 text-slate-600 text-sm font-semibold flex items-center justify-center`}
        >
          +{overflowCount}
        </div>
      )}
    </div>
  );
};

export default AvatarGroup;
