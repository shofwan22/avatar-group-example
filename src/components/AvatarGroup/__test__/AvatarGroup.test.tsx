import { render, screen, within } from '@testing-library/react';
import AvatarGroup from '../AvatarGroup';

describe('<AvatarGroup />', () => {
  const users = [
    { name: 'Alice Doe', image: './ava.jpeg' },
    { name: 'Bob Marley', image: './ava.jpeg' },
    { name: 'Charlie Van', image: '' },
    { name: 'Dave Roney', image: './ava.jpeg' },
    { name: 'De Jong', image: './ava.jpeg' },
    { name: 'Frank Lamp', image: './ava.jpeg' },
  ];
  const maxLength = 4;
  const size = 'md';

  it('should render correctly', () => {
    const { container } = render(
      <AvatarGroup users={users} maxLength={maxLength} size={size} />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders correct number of avatars', () => {
    const { container } = render(
      <AvatarGroup users={users} maxLength={maxLength} size={size} />
    );
    const avatarContainer = screen.getByTestId('avatar-container');
    const avatarComponents = screen.getAllByTestId('avatar-component');
    expect(avatarContainer).toBeInTheDocument();
    expect(avatarComponents).toHaveLength(maxLength);

    expect(container).toMatchSnapshot();
  });

  it('should render the overflow count correctly', () => {
    const { container } = render(
      <AvatarGroup users={users} maxLength={maxLength} size={size} />
    );
    const avatarMore = screen.getByTestId('avatar-more');

    expect(avatarMore).toBeInTheDocument();
    expect(avatarMore).toHaveTextContent('+2');

    expect(container).toMatchSnapshot();
  });

  it('should render name initial if there is no image', () => {
    const { container } = render(
      <AvatarGroup users={users} maxLength={maxLength} size={size} />
    );
    const avatarContainer = screen.getByTestId('avatar-container');
    const userInitial = 'CV'; //  { name: 'Charlie Van', image: '' }
    expect(within(avatarContainer).getByText(userInitial)).toBeTruthy();

    expect(container).toMatchSnapshot();
  });

  it('should render with different size', () => {
    const { container } = render(
      <AvatarGroup users={users} maxLength={maxLength} size="lg" />
    );

    const avatarContainer = screen.getByTestId('avatar-container');
    const avatarComponents = screen.getAllByTestId('avatar-component');
    const avatarMore = screen.getByTestId('avatar-more');

    expect(avatarContainer).toBeInTheDocument();
    expect(avatarComponents).toHaveLength(maxLength);
    expect(avatarMore).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('should render with different max length', () => {
    const { container } = render(
      <AvatarGroup users={users} maxLength={2} size={size} />
    );

    const avatarComponents = screen.getAllByTestId('avatar-component');

    expect(avatarComponents).toHaveLength(2);

    expect(container).toMatchSnapshot();
  });
});
