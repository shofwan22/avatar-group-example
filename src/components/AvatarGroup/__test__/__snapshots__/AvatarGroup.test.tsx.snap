// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AvatarGroup /> renders correct number of avatars 1`] = `
<div>
  <div
    class="flex items-center -space-x-4"
    data-testid="avatar-container"
  >
    <div
      data-testid="avatar-component"
    >
      <img
        alt="avatar"
        class="h-10 w-10 rounded-full"
        data-testid="avatar-image"
        src="./ava.jpeg"
      />
    </div>
    <div
      data-testid="avatar-component"
    >
      <img
        alt="avatar"
        class="h-10 w-10 rounded-full"
        data-testid="avatar-image"
        src="./ava.jpeg"
      />
    </div>
    <div
      data-testid="avatar-component"
    >
      <div
        class="flex-shrink-0 h-10 w-10 rounded-full border-2 border-white bg-green-400 text-white text-base font-semibold flex items-center justify-center"
        data-testid="avatar-name"
      >
        CV
      </div>
    </div>
    <div
      data-testid="avatar-component"
    >
      <img
        alt="avatar"
        class="h-10 w-10 rounded-full"
        data-testid="avatar-image"
        src="./ava.jpeg"
      />
    </div>
    <div
      class="flex-shrink-0 h-10 w-10 rounded-full border-2 border-white bg-slate-300 text-slate-600 text-sm font-semibold flex items-center justify-center"
      data-testid="avatar-more"
    >
      +
      2
    </div>
  </div>
</div>
`;

exports[`<AvatarGroup /> should render correctly 1`] = `
<div>
  <div
    class="flex items-center -space-x-4"
    data-testid="avatar-container"
  >
    <div
      data-testid="avatar-component"
    >
      <img
        alt="avatar"
        class="h-10 w-10 rounded-full"
        data-testid="avatar-image"
        src="./ava.jpeg"
      />
    </div>
    <div
      data-testid="avatar-component"
    >
      <img
        alt="avatar"
        class="h-10 w-10 rounded-full"
        data-testid="avatar-image"
        src="./ava.jpeg"
      />
    </div>
    <div
      data-testid="avatar-component"
    >
      <div
        class="flex-shrink-0 h-10 w-10 rounded-full border-2 border-white bg-green-400 text-white text-base font-semibold flex items-center justify-center"
        data-testid="avatar-name"
      >
        CV
      </div>
    </div>
    <div
      data-testid="avatar-component"
    >
      <img
        alt="avatar"
        class="h-10 w-10 rounded-full"
        data-testid="avatar-image"
        src="./ava.jpeg"
      />
    </div>
    <div
      class="flex-shrink-0 h-10 w-10 rounded-full border-2 border-white bg-slate-300 text-slate-600 text-sm font-semibold flex items-center justify-center"
      data-testid="avatar-more"
    >
      +
      2
    </div>
  </div>
</div>
`;

exports[`<AvatarGroup /> should render name initial if there is no image 1`] = `
<div>
  <div
    class="flex items-center -space-x-4"
    data-testid="avatar-container"
  >
    <div
      data-testid="avatar-component"
    >
      <img
        alt="avatar"
        class="h-10 w-10 rounded-full"
        data-testid="avatar-image"
        src="./ava.jpeg"
      />
    </div>
    <div
      data-testid="avatar-component"
    >
      <img
        alt="avatar"
        class="h-10 w-10 rounded-full"
        data-testid="avatar-image"
        src="./ava.jpeg"
      />
    </div>
    <div
      data-testid="avatar-component"
    >
      <div
        class="flex-shrink-0 h-10 w-10 rounded-full border-2 border-white bg-green-400 text-white text-base font-semibold flex items-center justify-center"
        data-testid="avatar-name"
      >
        CV
      </div>
    </div>
    <div
      data-testid="avatar-component"
    >
      <img
        alt="avatar"
        class="h-10 w-10 rounded-full"
        data-testid="avatar-image"
        src="./ava.jpeg"
      />
    </div>
    <div
      class="flex-shrink-0 h-10 w-10 rounded-full border-2 border-white bg-slate-300 text-slate-600 text-sm font-semibold flex items-center justify-center"
      data-testid="avatar-more"
    >
      +
      2
    </div>
  </div>
</div>
`;

exports[`<AvatarGroup /> should render the overflow count correctly 1`] = `
<div>
  <div
    class="flex items-center -space-x-4"
    data-testid="avatar-container"
  >
    <div
      data-testid="avatar-component"
    >
      <img
        alt="avatar"
        class="h-10 w-10 rounded-full"
        data-testid="avatar-image"
        src="./ava.jpeg"
      />
    </div>
    <div
      data-testid="avatar-component"
    >
      <img
        alt="avatar"
        class="h-10 w-10 rounded-full"
        data-testid="avatar-image"
        src="./ava.jpeg"
      />
    </div>
    <div
      data-testid="avatar-component"
    >
      <div
        class="flex-shrink-0 h-10 w-10 rounded-full border-2 border-white bg-green-400 text-white text-base font-semibold flex items-center justify-center"
        data-testid="avatar-name"
      >
        CV
      </div>
    </div>
    <div
      data-testid="avatar-component"
    >
      <img
        alt="avatar"
        class="h-10 w-10 rounded-full"
        data-testid="avatar-image"
        src="./ava.jpeg"
      />
    </div>
    <div
      class="flex-shrink-0 h-10 w-10 rounded-full border-2 border-white bg-slate-300 text-slate-600 text-sm font-semibold flex items-center justify-center"
      data-testid="avatar-more"
    >
      +
      2
    </div>
  </div>
</div>
`;

exports[`<AvatarGroup /> should render with different max length 1`] = `
<div>
  <div
    class="flex items-center -space-x-4"
    data-testid="avatar-container"
  >
    <div
      data-testid="avatar-component"
    >
      <img
        alt="avatar"
        class="h-10 w-10 rounded-full"
        data-testid="avatar-image"
        src="./ava.jpeg"
      />
    </div>
    <div
      data-testid="avatar-component"
    >
      <img
        alt="avatar"
        class="h-10 w-10 rounded-full"
        data-testid="avatar-image"
        src="./ava.jpeg"
      />
    </div>
    <div
      class="flex-shrink-0 h-10 w-10 rounded-full border-2 border-white bg-slate-300 text-slate-600 text-sm font-semibold flex items-center justify-center"
      data-testid="avatar-more"
    >
      +
      4
    </div>
  </div>
</div>
`;

exports[`<AvatarGroup /> should render with different size 1`] = `
<div>
  <div
    class="flex items-center -space-x-4"
    data-testid="avatar-container"
  >
    <div
      data-testid="avatar-component"
    >
      <img
        alt="avatar"
        class="h-20 w-20 rounded-full"
        data-testid="avatar-image"
        src="./ava.jpeg"
      />
    </div>
    <div
      data-testid="avatar-component"
    >
      <img
        alt="avatar"
        class="h-20 w-20 rounded-full"
        data-testid="avatar-image"
        src="./ava.jpeg"
      />
    </div>
    <div
      data-testid="avatar-component"
    >
      <div
        class="flex-shrink-0 h-20 w-20 rounded-full border-2 border-white bg-green-400 text-white text-lg font-semibold flex items-center justify-center"
        data-testid="avatar-name"
      >
        CV
      </div>
    </div>
    <div
      data-testid="avatar-component"
    >
      <img
        alt="avatar"
        class="h-20 w-20 rounded-full"
        data-testid="avatar-image"
        src="./ava.jpeg"
      />
    </div>
    <div
      class="flex-shrink-0 h-20 w-20 rounded-full border-2 border-white bg-slate-300 text-slate-600 text-sm font-semibold flex items-center justify-center"
      data-testid="avatar-more"
    >
      +
      2
    </div>
  </div>
</div>
`;
